tf_quant_finance/models/heston/approximations/asian_prices.py

```diff
@@ -0,0 +1,509 @@
+    theta = tf.convert_to_tensor(theta, dtype=dtype, name='theta')
+    theta = tf.broadcast_to(theta, strikes.shape)
+    volvol = tf.convert_to_tensor(volvol, dtype=dtype, name='volvol')
+    volvol = tf.broadcast_to(volvol, strikes.shape)
+    rho = tf.convert_to_tensor(rho, dtype=dtype, name='rho')
+    rho = tf.broadcast_to(rho, strikes.shape)
+
+    if discount_rates is not None:
+      discount_rates = tf.convert_to_tensor(
+          discount_rates, dtype=dtype, name='discount_rates')
+      discount_factors = tf.exp(-discount_rates * expiries)
+    elif discount_factors is not None:
+      discount_factors = tf.convert_to_tensor(
+          discount_factors, dtype=dtype, name='discount_factors')
+      discount_rates = -tf.math.log(discount_factors) / expiries
+    else:
+      discount_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='discount_rates')
+      discount_factors = tf.convert_to_tensor(
+          1.0, dtype=dtype, name='discount_factors')
+
+    if dividend_rates is None:
+      dividend_rates = tf.convert_to_tensor(
+          0.0, dtype=dtype, name='dividend_rates')
+
+    if forwards is not None:
+      forwards = tf.convert_to_tensor(forwards, dtype=dtype, name='forwards')
+      spots = forwards * tf.exp(-(discount_rates - dividend_rates) * expiries)
+    else:
+      spots = tf.convert_to_tensor(spots, dtype=dtype, name='spots')
+      forwards = spots * tf.exp((discount_rates - dividend_rates) * expiries)
+
+    forwards = tf.broadcast_to(forwards, strikes.shape)
+    spots = tf.broadcast_to(spots, strikes.shape)
+
+    #  To account for seasoning, we keep pricing time fixed at t=0 and adjust
+    #  the pricing parameters as follows:
+    #    - calculate prefactor running_prod ^ (1/#_total_fixings)
+    #    - strikes \to strikes / prefactor
+    #    - k_star \to #_past_fixings
+    #    - prepend #_past_fixings negative values to sampling times
+    #    - prices \to prefactor * prices
+    #
+    #  This is slightly different to the approach in the paper, where instead
+    #  t is allowed to increase above 0 as pricing time moves forwards
+
+    if past_fixings is None:
+      running_accumulator = tf.convert_to_tensor(1.0, dtype=dtype)
+      fixing_count = 0
+    else:
+      running_accumulator = tf.reduce_prod(past_fixings, 0)
+      fixing_count = past_fixings.shape[0]
+
+    start_times = tf.zeros(spots.shape[0], dtype=dtype)
+    tau_k = tf.concat([tf.expand_dims(start_times, axis=0),
+                      sampling_times,
+                      tf.expand_dims(expiries, axis=0)], 0)
+
+    n = sampling_times.shape[0] + fixing_count
+    prefactor = tf.math.pow(running_accumulator, 1 / n)
+    adj_strikes = strikes / prefactor
+
+    sampling_times = tf.concat(
+      [
+        tf.constant(
+          [[-1] * sampling_times.shape[1]] * fixing_count,
+          dtype=dtype,
+          shape=(fixing_count, sampling_times.shape[1])
+        ),
```

Comment:
also, is this the same as 
`-1 * tf.ones([fixing_count, utils.get_shape(sampling_times)[1]],
                      dtype=dtype)`?

Reply:
Great spot, it is exactly. Updated, along with a comment on shape.